### PR TITLE
Add telegram bot to chat with roborock

### PR DESCRIPTION
**Pull request by thejoe8495**
I added the telegram bot API to talk with Roborock, you can start, stop. On every status change the Roborock send you an message over telegram.

### CODE REVIEW

**Comment by bzz11g** [comment]
wow, this will make it possible to completely replace the standard application, and send the robot for cleaning without being on the same network, I hope that this will be added in the firmware with the form in the settings where you can enter the http token of your bot in easy way.

File: client/settings-telegrambot.html [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+        var settingsTelegramclients = document.getElementById('settings-telegram-clients');
+
+        ons.getScriptPage().onShow = function () {
+            getSettingsTelegramPage();
+        };
+        document.querySelector('ons-switch').addEventListener('change', setSettingsTelegramPage);
+        document.querySelector('ons-input').addEventListener('change', setSettingsTelegramPage);
+        function getSettingsTelegramPage() {
+            loadingBarSettingsTelegram.setAttribute("indeterminate", "indeterminate");
+            fn.request("api/get_telegram_config", "GET", function (err, res) {
```
Comment (rand256) [inline client/settings-telegrambot.html]: Generally this is deprecated, should use fn.prequest with promises.

File: lib/Telegrambot.js [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+    this.running = 0;
+    this.events = options.events;
+    this.vacuum = options.vacuum;
+    this.configuration = options.configuration;
+    this.initiate(options);
+    this.laststate = {state: 0, error:0}
+    this.shutup = false;
+
+    this.events.on("miio.bin_full", (statusData) => {
+        // console.log(statusData);
```
Comment (rand256) [inline lib/Telegrambot.js:27]: Is everything here really should be commented out?

File: client/settings-telegrambot.html [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+                            "</ons-list-item>"
+                        );
+
+                        elem.querySelector('.client-delete').addEventListener('click', function (event) {
+                            deleteClient(client.id);
+                        });
+                        settingsTelegramclients.appendChild(elem);
+                    });
+                } else {
+                    ons.notification.toast(err, { buttonLabel: 'Dismiss', timeout: window.fn.toastErrorTimeout })
```
Comment (rand256) [inline client/settings-telegrambot.html]: There's now a shortcut functions for generic toasts: fn.notificationToastOK and fn.notificationToastError.

File: lib/Telegrambot.js [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+        //     if(!self.shutup)
+        //         self.slimbot.sendMessage(client.id, statusData);
+        // });
+    });
+    this.events.on("miio.status", (statusData) => {
+        if ((self.laststate.state == statusData.state && self.laststate.error == statusData.error_code) || self.shutup) return;
+        res = Vacuum.GET_STATE_CODE_DESCRIPTION(statusData.state);
+        if (statusData.error_code != 0)
+            res += " " + Vacuum.GET_ERROR_CODE_DESCRIPTION(statusData.error_code);
+        if (self.slimbot == null) return;
```
Comment (rand256) [inline lib/Telegrambot.js:39]: Is there a reason to place this here and not at the beginning?

File: lib/Telegrambot.js [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+        // console.log(statusData);
+        // if (!this.slimbot) return;
+        // self.clients.forEach(function (client) {
+        //     if(!self.shutup)
+        //         self.slimbot.sendMessage(client.id, statusData);
+        // });
+    });
+    this.events.on("miio.status", (statusData) => {
+        if ((self.laststate.state == statusData.state && self.laststate.error == statusData.error_code) || self.shutup) return;
+        res = Vacuum.GET_STATE_CODE_DESCRIPTION(statusData.state);
```
Comment (rand256) [inline lib/Telegrambot.js]: Is `res` declared somewhere?

File: lib/Valetudo.js [hunk 6]
```diff
@@ -48,6 +49,30 @@ const Valetudo = function() {
         vacuum: this.vacuum,
     });
 
+    if(this.configuration.get("telegram_token")) {
+        this.telegramClient = new Telegrambot({ 
+            vacuum: this.vacuum, 
+            events: this.events, 
+            configuration: this.configuration, 
+            clients: this.configuration.get("telegram_clients"),
```
Comment (rand256) [inline lib/Valetudo.js]: Why this and following variables set here and not inside Telegrambot constructor?

File: lib/Valetudo.js [hunk 7]
```diff
@@ -48,6 +49,30 @@ const Valetudo = function() {
...
+            vacuum: this.vacuum, 
+            events: this.events, 
+            configuration: this.configuration, 
+            clients: this.configuration.get("telegram_clients"), 
+            telegramtoken: this.configuration.get("telegram_token"), 
+            telegrampassword: this.configuration.get("telegram_password"), 
+            sendconsumables: this.configuration.get("telegram_send_consumables"), 
+            sendconsumablesevery: this.configuration.get("telegram_send_consumables_every") 
+        }); 
+    } else {
```
Comment (rand256) [inline lib/Valetudo.js:58]: Can't see the reason for this construct here at all.

File: lib/Valetudo.js [hunk 8]
```diff
@@ -48,6 +49,30 @@ const Valetudo = function() {
         vacuum: this.vacuum,
     });
 
+    if(this.configuration.get("telegram_token")) {
+        this.telegramClient = new Telegrambot({
```
Comment (rand256) [inline lib/Valetudo.js:53]: I believe it should be better to be able to completely disable Telegrambot in config.json for users who never use telegram, so this would never be called for them.

File: lib/miio/Dummycloud.js [hunk 9]
```diff
@@ -97,7 +97,7 @@ const Dummycloud = function (options) {
                         break;
                     case "props":
                         Object.assign(this.connectedRobot.status, decodedResponse.msg.params);
-                        this.events.emit("miio.status", this.connectedRobot.status);
+                        //this.events.emit("miio.status", this.connectedRobot.status);
```
Comment (rand256) [inline lib/miio/Dummycloud.js:100]: So, just comment out and that's it?

File: lib/miio/Dummycloud.js [hunk 10]
```diff
@@ -150,6 +150,15 @@ const Dummycloud = function (options) {
 
                         break;
                     case "event.back_to_dock": //TODO
+                        this.events.emit("miio.home", true);
+                        break;
```
Comment (rand256) [inline lib/miio/Dummycloud.js:154]: This will be left without responseJSON then. Is it intended?

File: lib/miio/Vacuum.js [hunk 11]
```diff
@@ -88,7 +88,7 @@ Vacuum.prototype.sendMessage = function(method, args, options, callback) {
         // since the service automatically restarts, just let it do it
         // but only if there were successful communications before (hello ./develop/run)
         console.error('failed to get handshake for message:', method, args, options);
-        throw new Error("Unable to reach vacuum, handshake failed");
+        throw new Error("Unable to reach vacuum, handshake failed is the token right?");
```
Comment (rand256) [inline lib/miio/Vacuum.js:91]: Doesn't seem to be related to telegrambot. But ok.

File: lib/webserver/WebServer.js [hunk 12]
```diff
@@ -162,6 +163,69 @@ const WebServer = function (options) {
...
     });
 
+    // Telegram settings
+    this.app.get("/api/get_telegram_config", function (req, res) {
+        if(self.configuration.get("telegram_token")){
+            token = self.configuration.get("telegram_token");
+            password = self.configuration.get("telegram_password");
+            clients = self.configuration.get("telegram_clients");
+            telegram_send_consumables = self.configuration.get("telegram_send_consumables");
+            telegram_send_consumables_every = self.configuration.get("telegram_send_consumables_every");
```
Comment (rand256) [inline lib/webserver/WebServer.js:178]: Can't see where all these variables were declared before use.

File: lib/webserver/WebServer.js [hunk 13]
```diff
@@ -162,6 +163,69 @@ const WebServer = function (options) {
...
+                sendconsumables: false,
+                sendconsumablesevery: false
+            });
+        }
+        return res.json({res:"success"});
+    });
+
+    this.app.delete("/api/telegram_client/:clientID", function (req, res) {
+        clients = self.configuration.get("telegram_clients");
+        newclients = [];
```
Comment (rand256) [inline lib/webserver/WebServer.js]: Same here.

**Review by rand256 — COMMENTED** [review]
Maybe it could be useful for those who use telegram, but it needs to have an option to get it completely disabled for those who don't, so it won't waste memory. Also quiet a lot of work is to be done on the code to be merged (localization support at least).

I'll take a look on it later if nobody would do it earlier.

File: lib/Telegrambot.js (hunk 2 above)
Comment (thejoe8495) [reply lib/Telegrambot.js:27]: Yes my bot start every day at 10 am, befor the robot start it send me one massage with bin full, after the start it send me additional 5 messages. At the moment i don't know why. But i will find it out.

File: client/settings-telegrambot.html (hunk 1 above)
Comment (thejoe8495) [reply client/settings-telegrambot.html]: I can change it, i use the same coding style who use Hypfer

File: lib/Telegrambot.js (hunk 4 above)
Comment (thejoe8495) [reply lib/Telegrambot.js:39]: It's a good point. i move it

File: lib/Valetudo.js (hunk 8 above)
Comment (thejoe8495) [reply lib/Valetudo.js:53]: Hmmm yes if there is no telegram_token set telegram will be ignored. You need to set telegram_token on the settingspage. Do you have another idea?

File: lib/miio/Dummycloud.js (hunk 9 above)
Comment (thejoe8495) [reply lib/miio/Dummycloud.js:100]: Props send you every 30 minutes the status of the robot, i can't see any reason why i need this. There is no change in this status messages. When the robot start the cleaning, error, etc. it sends `miio.status`

File: lib/webserver/WebServer.js (hunk 12 above)
Comment (thejoe8495) [reply lib/webserver/WebServer.js:178]: Yes it's my first nodejs program ;). I will make it better

File: lib/Valetudo.js (hunk 6 above)
Comment (thejoe8495) [reply lib/Valetudo.js]: I need the variables vacuum, events, ... i have no idea how i can access this variables inside the telegramobject. I add the configuration later, so i don't need to transfer the variables like telegram_token. Is there a another way to access vaccum or events?

Comment (rand256) [reply lib/Valetudo.js]: That was comment about passing some specific configuration properties along with configuration object itself. Vacuum and events were quoted here because github shows some context before the commented line it seems.

File: lib/Valetudo.js (hunk 8 above)
Comment (rand256) [reply lib/Valetudo.js:53]: Current code creates `Telegrambot` object regardless of `telegram_token` being set or not, that I was talking about. In my country, for example, telegram is forbidden by the law. I can't directly use it. Why should I have its components loaded into valetudo so they would consume limited vacuum's RAM for nothing? There should be an ability to completely disable it when it's not needed.

**Comment by thejoe8495** [comment]
@rand256 is there function to get the translated strings in code behind? I want to use it in the telegram.js

**Comment by rand256** [comment]
Translation is done via i18next.js with ​loc-i18next plugin. See examples in any page inside ./client.

**Comment by thejoe8495** [comment]
I hope i find all places. The translation in telegram.js came in the next version :sweat_smile:

**Comment by MadJoker0815** [comment]
I created a version of the German translation including the telegram add on.
https://github.com/MadJoker0815/valetudo/blob/MadJoker0815-patch-1-w-telegram/client/locales/de.json

**Comment by rand256** [comment]
Thanks. This pull request has been integrated manually.

**Comment by rand256** [comment]
@MadJoker0815, could you please update your German translation to current string set?

**Comment by MadJoker0815** [comment]
Translation is updated. Hope it fits.

**Comment by rand256** [comment]
Committed, thank you.
